File: native-micronaut-api/src/main/java/com/xavierbouclet/nativeapi/MovieService.java
package com.xavierbouclet.nativeapi;

import jakarta.inject.Inject;
import jakarta.inject.Singleton;
import java.util.*;

@Singleton
record MovieService(
    @Inject MovieRepository repository,
    @Inject ActorRepository actorRepository
) {

    public Iterable<Movie> list() {
        return repository.findAll();
    }

//    public list(title: String): List<Movie> {
//        return repository.findByTitle(title)
//    }

    public Movie add(Movie movie) {
        return repository.save(movie);
    }

    public Movie modify(Movie movie) {
        return repository.save(movie);
    }

    public UUID delete(UUID id) {
        repository.deleteById(id);
        return id;
    }

    public Movie addActorToMovie(UUID id, UUID idActor) {
        var movie = repository.findById(id).get();
        var actor = actorRepository.findById(idActor).get();
        movie.getActors().add(actor);
        actor.getMovies().add(movie);
        return repository.save(movie);
    }

    public Movie get(UUID id) {
        return repository.findById(id).get();
    }

    public Movie removeActorFromMovie(UUID id, UUID idActor) {
        var movie = repository.findById(id).get();
        var actor = actorRepository.findById(idActor).get();
        movie.getActors().remove(actor);
        actor.getMovies().remove(movie);
        return repository.save(movie);
    }

}

File: native-micronaut-api/src/main/kotlin/com/xavierbouclet/nativeapi/MovieService.kt
package com.xavierbouclet.nativeapi

import jakarta.inject.Inject
import jakarta.inject.Singleton
import java.util.*

@Singleton
class MovieService(
    @Inject val repository: MovieRepository,
    @Inject val actorRepository: ActorRepository
) {

    fun list(): MutableIterable<Movie> {
        return repository.findAll()
    }

//    fun list(title: String): List<Movie> {
//        return repository.findByTitle(title)
//    }

    fun add(movie: Movie): Movie {
        return repository.save(movie)
    }

    fun modify(movie: Movie): Movie {
        return repository.save(movie)
    }

    fun delete(id: UUID): UUID {
        repository.deleteById(id)
        return id
    }

    fun addActorToMovie(id: UUID, idActor: UUID): Movie {
        val movie = repository.findById(id).get()
        val actor = actorRepository.findById(idActor).get()
        movie.actors?.add(actor)
        actor.movies?.add(movie)
        return repository.save(movie)
    }

    fun get(id: UUID): Movie? {
        return repository.findById(id).orElseGet(null)
    }

    fun removeActorFromMovie(id: UUID, idActor: UUID): Movie {
        val movie = repository.findById(id).get()
        val actor = actorRepository.findById(idActor).get()
        movie.actors?.remove(actor)
        actor.movies?.remove(movie)
        return repository.save(movie)
    }

}
